rust: Always inject vscrt args

xclaesse opened the pull request:
Rustc assumes "md" by default, but we should always respect Meson
    options. Building some targets with different vscrt makes them fail to
    link together. This is also what other compilers does.

xclaesse commented:
This is WIP because I don't think forcing a CRT on Rustc ever worked: https://github.com/mesonbuild/meson/pull/12367#issuecomment-1760614288. I'm expecting the CI to fail here.

xclaesse commented:
@dcbaker You did not make a unit test in https://github.com/mesonbuild/meson/pull/11902 and I have no idea how to reproduce. Any idea?

dcbaker commented:
I think the answer is that static libraries shouldn't be getting CRT arguments. In Mesa we only link Rust, C, and C++ together into shared libraries, so that would be the only tested path. I looked at the documentation for MSVC's `lib` command and don't see any arguments for setting the CRT, so I think this only should be done for dynamic objects.

dcbaker commented on mesonbuild/compilers/compilers.py:
```diff
@@ -325,7 +327,7 @@ class CompileCheckMode(enum.Enum):
     OptionKey('b_bitcode'): coredata.UserBooleanOption('Generate and embed bitcode (only macOS/iOS/tvOS)', False),
     OptionKey('b_vscrt'): coredata.UserComboOption('VS run-time library type to use.',
-                                                   ['none', 'md', 'mdd', 'mt', 'mtd', 'from_buildtype', 'static_from_buildtype'],
+                                                   list(MSCRT_VALS) + ['from_buildtype', 'static_from_buildtype'],
```
You either need `sorted(MSVCRT_VALS)` here, or to use an `OrderedSet` for this, otherwise the `meson configure` output may be re-ordered

xclaesse replied:
Good catch. Made it a list, for small set it does not really make a difference.

dcbaker commented on mesonbuild/compilers/rust.py:
```diff
@@ -55,6 +55,14 @@ class RustCompiler(Compiler):
         '3': ['-W', 'warnings'],
     }
 
+    mscrt_args = {
```
Can we make this: `MSVCRT_ARGS: T.Mapping[str, T.List[str]] = ...`

All caps is PEP8 convention for "constant data", and making the type a `mapping` will let mypy yell at us if someone tries to mutate this. I would love to make this `ImmutableListProtocol`, but I think that may be an invasive change, and it's not in scope for this PR.

xclaesse replied:
done

dcbaker reviewed (approved):
Both of these patches are great cleanups. I have two small nits, but otherwise consider this RB

xclaesse commented:
> That series does have a unit test, I added a test.json to "20 rust and cpp" to test the various crt options. But It only exercises dynamic linking path, which is part of why I think we don't actually want to inject crt arguments into rlib and archive generation

You're right, it only makes sense for dll/exe. Updated this PR to reflect that.

xclaesse commented:
Hm, seems there is still something I misunderstood:

https://dev.azure.com/jussi0947/jussi/_build/results?buildId=19504&view=logs&j=5015ea34-1c35-577b-b206-c8bea8465a58&t=00c5cf7c-b41c-5970-0bf5-8751e484378e&l=2957
